components/authz-service/storage/postgres/migration/sql/44_policy_rules.up.sql

```diff
@@ -0,0 +1,71 @@
+BEGIN;
+
+CREATE TYPE iam_rule_type AS ENUM ('node', 'event');
+CREATE TYPE iam_attributes AS ENUM ('organization', 'chef-server',
+  'environment', 'role', 'chef-tag', 'policy-name', 'policy-group');
+CREATE TYPE iam_condition_operator AS ENUM ('equals', 'member-of');
+
+CREATE TABLE iam_project_rules (
+  db_id SERIAL PRIMARY KEY,
+  id TEXT NOT NULL UNIQUE,
+  project_id TEXT REFERENCES iam_projects ON DELETE CASCADE,
+  name TEXT NOT NULL,
+  type iam_rule_type NOT NULL
+);
+
+CREATE TABLE iam_rule_conditions (
+  db_id SERIAL PRIMARY KEY,
+  rule_id INTEGER REFERENCES iam_project_rules ON DELETE CASCADE,
+  value TEXT NOT NULL,
+  attribute iam_attributes NOT NULL,
+  operator iam_condition_operator NOT NULL
+);
+
+CREATE OR REPLACE FUNCTION
+  projects_match_for_rule(_project_id TEXT, _projects_filter TEXT[])
```

Comment:
Function already exists as `projects_match`, so delete this function and just use `projects_match` in the query functions; see my next comment.

Reply:
They are a bit different:

```
CREATE OR REPLACE FUNCTION
  projects_match(_projects TEXT[], _projects_filter TEXT[])
  RETURNS BOOLEAN AS $$
    BEGIN
      RETURN (
        -- no projects filter requested (length 0) will be the case for v2.0 or v2.1 ["*"]
        array_length(_projects_filter, 1) IS NULL
        -- projects filter intersects with projects for row
        OR _projects && _projects_filter
        -- projects for row is an empty array, check if (unassigned) in project filter
        OR (array_length(_projects, 1) IS NULL AND '{(unassigned)}' && _projects_filter)
      );
    END
$$ LANGUAGE PLPGSQL;
```

```
CREATE OR REPLACE FUNCTION
  projects_match_for_rule(_project_id TEXT, _projects_filter TEXT[])
  RETURNS BOOLEAN AS $$
    BEGIN
      RETURN (
        array_length(_projects_filter, 1) IS NULL
        OR _project_id = ANY (_projects_filter)
      );
    END
$$ LANGUAGE PLPGSQL;
```

Rules can only have a single project which will not be in an array already like other objects. It will also never be unassigned so we don't wanna go down that logic path. Rules will always have a one and only one project (which in fact cannot change).